File: eval.ml
open Printf

open Ast
open Name
open Names
open Value
open Parse

let rec eval env = function
  | AST.Name n -> !(try NameMap.find n !env
                    with Not_found ->
                      Value.fail Names.e_name (Name.to_string n))
  | AST.Symbol n -> Value.Symbol n
  | AST.String s -> Value.String s
  | AST.Int i -> Value.Int i
  | AST.Float f -> Value.Float f
  | AST.Char c -> Value.Char c
  | AST.List l -> Value.List (List.map (eval env) l)
  | AST.Record r -> Value.Record (Names.record, NameMap.map (eval env) r)
  | AST.Do [] -> Value.None
  | AST.Do [expr] -> eval env expr
  | AST.Do (expr :: exprs) -> let _ = eval env expr in eval env (AST.Do exprs)
  | AST.Def (n, v) ->
      env := NameMap.add n (ref (eval env v)) !env;
      Value.None
  | AST.Set (n, v) ->
      (try (NameMap.find n !env) := (eval env v)
       with Not_found ->
         Value.fail
           Names.e_name ("name " ^ (Name.to_string n) ^ " not found"));
      Value.None
  | AST.Undef n ->
      env := NameMap.remove n !env;
      Value.None
  | AST.Let (bindings, a) ->
      let new_env =
        List.fold_left
          (fun a (n, v) ->
             NameMap.add n (ref (eval env v)) a) !env bindings in
      eval (ref new_env) a
  | AST.If [] -> Value.None
  | AST.If (elseexpr :: []) -> eval env elseexpr
  | AST.If (ifcond :: ifexpr :: rest) ->
    (if Value.to_bool (eval env ifcond)
     then eval env ifexpr
     else eval env (AST.If rest))
  | AST.App (f, a) ->
      (match (eval env f) with
         | Value.Function f' -> f' (eval env a)
         | Value.String s ->
             (match (eval env a) with
                | Value.Int i ->
                    (try Value.Char s.[i]
                     with Invalid_argument "index out of bounds" ->
                       Value.fail_value Names.e_index_nf (Value.Int i))
                | _ -> Value.fail Names.e_type "can't apply non-integer to string")
         | Value.List l ->
             (match (eval env a) with
                | Value.Int i -> (try List.nth l i
                                  with Failure "nth" ->
                                    Value.fail_value
                                      Names.e_index_nf (Value.Int i))
                | _ -> Value.fail Names.e_type "can't apply non-integer to list")
         | Value.Record (tag, r) as r' ->
             (match (eval env a) with
                | Value.Symbol n -> (try NameMap.find n r
                                     with Not_found ->
                                       Value.fail_value
                                         Names.e_field_nf (Value.Symbol n))
                | _ as arg -> 
                    (try
                       (match Value.generic (Name.of_string "call") tag r' with
                          | Value.Function f -> f arg
                          | _ -> Value.fail Names.e_type "non-function in generics table")
                     with Not_found ->
                       Value.fail Names.e_type ("call is not defined for " ^ (Name.to_string tag))))
         | _ as v -> Value.fail Names.e_type ("call of non-function: " ^ (Value.to_string v)))
  | AST.Fun (AST.EmptyPat, ast) ->
      Value.Function
        (function
           | Value.None -> eval (ref !env) ast
           | _ -> Value.fail Names.e_type "function takes no arguments")
  | AST.Fun (AST.ScalarPat name, ast) ->
      Value.Function (fun arg ->
                        eval (ref (NameMap.add name (ref arg) !env)) ast)
  | AST.Fun (AST.ListPat names, ast) ->
      Value.Function
        (function
           | Value.List args ->
               let new_env =
                 (try
                    List.fold_left2
                      (fun a n v -> NameMap.add n (ref v) a) !env names args
                  with
                    | Invalid_argument "List.fold_left2" ->
                        Value.fail
                          Names.e_type
                          (sprintf
                             "function requires a list of size %d"
                             (List.length names))) in
               eval (ref new_env) ast
           | _ -> Value.fail Names.e_type "function requires a list argument")
  | AST.Fun (AST.RecordPat names, ast) ->
      Value.Function
        (function
           | Value.Record (_, args) ->
               let new_env =
                 List.fold_left
                   (fun a n ->
                      try NameMap.add n (ref (NameMap.find n args)) a
                      with Not_found ->
                        Value.fail
                          Names.e_type
                          (sprintf
                             "function requires a record with field %s"
                             (Name.to_string n))) !env names in
               eval (ref new_env) ast
           | _ ->
               Value.fail Names.e_type "function requires a record argument")
  | AST.Try (expr, catches) ->
      (try (eval env expr)
       with Value.Error e as exn ->
         let rec loop = function
           | [] -> raise exn
           | (catch :: catches) ->
               (match catch with
                  | AST.ValCatch (n, a) ->
                      let new_env = NameMap.add n (ref e) !env in
                      (eval (ref new_env) a)
                  | AST.PatCatch (AST.EmptyPat, a)
                    -> (eval env a)
                  | AST.PatCatch (AST.ScalarPat n, a) ->
                      let new_env = NameMap.add n (ref e) !env in
                      (eval (ref new_env) a)
                  | AST.PatCatch (AST.ListPat names, a) ->
                      (match e with
                         | Value.List l ->
                             (try
                                (let new_env = List.fold_left2
                                   (fun a n v -> NameMap.add n (ref v) a)
                                   !env names l in
                                 (eval (ref new_env) a))
                              with Invalid_argument "List.fold_left2" ->
                                loop catches)
                         | _ -> loop catches)
                  | AST.PatCatch (AST.RecordPat names, a) ->
                      (match e with
                         | Value.Record (_, r) ->
                             (try
                                (let new_env =
                                   List.fold_left
                                     (fun a n ->
                                        NameMap.add
                                          n (ref (NameMap.find n r)) a)
                                     !env names in
                                 (eval (ref new_env) a))
                              with Not_found -> loop catches)
                         | _ -> loop catches))
         in loop catches)

let eval_stream env stream =
  let ast_stream = (parse_stream (lexer stream)) in
  let result = ref Value.None in
  Stream.iter (fun ast -> result := eval env ast) ast_stream;
  result

let eval_string env string =
  eval_stream env (Stream.of_string string)

File: rust/src/eval.rs
use std::rc::Rc;

use crate::ast::{Ast, Catch, Pattern};
use crate::env::{child_env, Env};
use crate::name::Name;
use crate::names::{NameMap, Names};
use crate::value::{SquareError, Value, GENERICS};

/// Evaluate an AST node in the given environment.
/// Directly mirrors the OCaml `eval` function.
pub fn eval(env: &Env, ast: &Ast) -> Result<Value, SquareError> {
    match ast {
        Ast::Name(n) => {
            let cell = env.borrow().get(*n).ok_or_else(|| {
                Value::error(Names::e_name(), &n.to_string())
            })?;
            Ok(cell.borrow().clone())
        }
        Ast::Symbol(n) => Ok(Value::Symbol(*n)),
        Ast::String(s) => Ok(Value::string(s.clone())),
        Ast::Int(i) => Ok(Value::Int(*i)),
        Ast::Float(f) => Ok(Value::Float(*f)),
        Ast::Char(c) => Ok(Value::Char(*c)),
        Ast::List(l) => {
            let vals: Result<Vec<Value>, _> = l.iter().map(|a| eval(env, a)).collect();
            Ok(Value::list(vals?))
        }
        Ast::Record(r) => {
            let mut map = NameMap::new();
            for (k, v) in r {
                map.insert(*k, eval(env, v)?);
            }
            Ok(Value::Record(Names::record(), Rc::new(map)))
        }
        Ast::Do(stmts) => {
            if stmts.is_empty() {
                return Ok(Value::None);
            }
            let mut result = Value::None;
            for stmt in stmts {
                result = eval(env, stmt)?;
            }
            Ok(result)
        }
        Ast::Def(n, v) => {
            let val = eval(env, v)?;
            env.borrow_mut().define(*n, val);
            Ok(Value::None)
        }
        Ast::Set(n, v) => {
            let val = eval(env, v)?;
            let cell = env.borrow().get(*n).ok_or_else(|| {
                Value::error(Names::e_name(), &n.to_string())
            })?;
            *cell.borrow_mut() = val;
            Ok(Value::None)
        }
        Ast::Undef(n) => {
            env.borrow_mut().remove(*n);
            Ok(Value::None)
        }
        Ast::Let(bindings, body) => {
            let new_env = child_env(env);
            for (n, v) in bindings {
                let val = eval(&new_env, v)?;
                new_env.borrow_mut().define(*n, val);
            }
            eval(&new_env, body)
        }
        Ast::If(clauses) => eval_if(env, clauses),
        Ast::App(f, a) => {
            let func_val = eval(env, f)?;
            eval_app(env, &func_val, a)
        }
        Ast::Fun(pat, body) => eval_fun(env, pat, body),
        Ast::Try(expr, catches) => eval_try(env, expr, catches),
    }
}

fn eval_if(env: &Env, clauses: &[Ast]) -> Result<Value, SquareError> {
    match clauses.len() {
        0 => Ok(Value::None),
        1 => eval(env, &clauses[0]),
        _ => {
            let cond = eval(env, &clauses[0])?;
            if cond.to_bool() {
                eval(env, &clauses[1])
            } else {
                eval_if(env, &clauses[2..])
            }
        }
    }
}

fn eval_app(env: &Env, func_val: &Value, arg_ast: &Ast) -> Result<Value, SquareError> {
    match func_val {
        Value::Function(f) => {
            let arg = eval(env, arg_ast)?;
            f(arg)
        }
        Value::String(s) => {
            let arg = eval(env, arg_ast)?;
            match arg {
                Value::Int(i) => {
                    let idx = if i < 0 { s.len() as i64 + i } else { i } as usize;
                    s.chars()
                        .nth(idx)
                        .map(Value::Char)
                        .ok_or_else(|| Value::error_value(Names::e_index_nf(), Value::Int(i)))
                }
                _ => Err(Value::error(
                    Names::e_type(),
                    "can't apply non-integer to string",
                )),
            }
        }
        Value::List(l) => {
            let arg = eval(env, arg_ast)?;
            match arg {
                Value::Int(i) => {
                    let idx = if i < 0 { l.len() as i64 + i } else { i } as usize;
                    l.get(idx)
                        .cloned()
                        .ok_or_else(|| Value::error_value(Names::e_index_nf(), Value::Int(i)))
                }
                _ => Err(Value::error(
                    Names::e_type(),
                    "can't apply non-integer to list",
                )),
            }
        }
        Value::Record(tag, r) => {
            let arg = eval(env, arg_ast)?;
            match &arg {
                Value::Symbol(n) => {
                    if let Some(v) = r.get(n) {
                        Ok(v.clone())
                    } else {
                        // Try generic call - clone out of borrow scope
                        let func = GENERICS.with(|g| {
                            let g = g.borrow();
                            g.get(Name::new("call"), *tag).cloned()
                        });
                        if let Some(Value::Function(f)) = func {
                            f(arg.clone())
                        } else {
                            Err(Value::error_value(
                                Names::e_field_nf(),
                                Value::Symbol(*n),
                            ))
                        }
                    }
                }
                _ => {
                    // Try generic call - clone out of borrow scope
                    let tag = *tag;
                    let func = GENERICS.with(|g| {
                        let g = g.borrow();
                        g.get(Name::new("call"), tag).cloned()
                    });
                    if let Some(Value::Function(f)) = func {
                        f(arg.clone())
                    } else {
                        Err(Value::error(
                            Names::e_type(),
                            &format!("call is not defined for {}", tag),
                        ))
                    }
                }
            }
        }
        _ => Err(Value::error(
            Names::e_type(),
            &format!(
                "call is not defined for {}",
                func_val.to_display_string()
            ),
        )),
    }
}

fn eval_fun(env: &Env, pat: &Pattern, body: &Ast) -> Result<Value, SquareError> {
    // Capture the env Rc directly (not a child copy). This mirrors OCaml's behavior
    // where closures capture the `ref` to the environment, so they see future `def`s.
    let captured_env = env.clone();
    let body = body.clone();
    let pat = pat.clone();

    match pat {
        Pattern::Empty => Ok(Value::Function(Rc::new(move |arg| {
            if let Value::None = arg {
                eval(&child_env(&captured_env), &body)
            } else {
                Err(Value::error(Names::e_type(), "function takes no arguments"))
            }
        }))),
        Pattern::Scalar(name) => Ok(Value::Function(Rc::new(move |arg| {
            let new_env = child_env(&captured_env);
            new_env.borrow_mut().define(name, arg);
            eval(&new_env, &body)
        }))),
        Pattern::List(names) => {
            let names = names.clone();
            Ok(Value::Function(Rc::new(move |arg| {
                if let Value::List(args) = arg {
                    if args.len() != names.len() {
                        return Err(Value::error(
                            Names::e_type(),
                            &format!(
                                "function requires a list of size {}",
                                names.len()
                            ),
                        ));
                    }
                    let new_env = child_env(&captured_env);
                    for (n, v) in names.iter().zip(args.iter().cloned()) {
                        new_env.borrow_mut().define(*n, v);
                    }
                    eval(&new_env, &body)
                } else {
                    Err(Value::error(
                        Names::e_type(),
                        "function requires a list argument",
                    ))
                }
            })))
        }
        Pattern::Record(names) => {
            let names = names.clone();
            Ok(Value::Function(Rc::new(move |arg| {
                if let Value::Record(_, ref r) = arg {
                    let new_env = child_env(&captured_env);
                    for n in &names {
                        let v = r.get(n).ok_or_else(|| {
                            Value::error(
                                Names::e_type(),
                                &format!(
                                    "function requires a record with field {}",
                                    n
                                ),
                            )
                        })?;
                        new_env.borrow_mut().define(*n, v.clone());
                    }
                    eval(&new_env, &body)
                } else {
                    Err(Value::error(
                        Names::e_type(),
                        "function requires a record argument",
                    ))
                }
            })))
        }
    }
}

fn eval_try(
    env: &Env,
    expr: &Ast,
    catches: &[Catch],
) -> Result<Value, SquareError> {
    match eval(env, expr) {
        Ok(v) => Ok(v),
        Err(err) => {
            let e = err.value;
            for catch in catches {
                match catch {
                    Catch::Val(n, a) => {
                        let new_env = child_env(env);
                        new_env.borrow_mut().define(*n, e.clone());
                        return eval(&new_env, a);
                    }
                    Catch::Pat(Pattern::Empty, a) => {
                        return eval(env, a);
                    }
                    Catch::Pat(Pattern::Scalar(n), a) => {
                        let new_env = child_env(env);
                        new_env.borrow_mut().define(*n, e.clone());
                        return eval(&new_env, a);
                    }
                    Catch::Pat(Pattern::List(names), a) => {
                        if let Value::List(ref l) = e {
                            if l.len() == names.len() {
                                let new_env = child_env(env);
                                for (n, v) in names.iter().zip(l.iter()) {
                                    new_env.borrow_mut().define(*n, v.clone());
                                }
                                return eval(&new_env, a);
                            }
                        }
                        // Pattern doesn't match, try next catch
                        continue;
                    }
                    Catch::Pat(Pattern::Record(names), a) => {
                        if let Value::Record(_, ref r) = e {
                            let mut all_found = true;
                            for n in names {
                                if !r.contains_key(n) {
                                    all_found = false;
                                    break;
                                }
                            }
                            if all_found {
                                let new_env = child_env(env);
                                for n in names {
                                    new_env
                                        .borrow_mut()
                                        .define(*n, r.get(n).unwrap().clone());
                                }
                                return eval(&new_env, a);
                            }
                        }
                        continue;
                    }
                }
            }
            // No catch matched, re-raise
            Err(SquareError::new(e))
        }
    }
}

/// Evaluate a string of Square source code in the given environment.
pub fn eval_string(env: &Env, source: &str) -> Result<Value, SquareError> {
    let mut lexer = crate::lexer::Lexer::new(source);
    let mut parser = crate::parser::Parser::new(&mut lexer);
    let program = parser.parse_program().map_err(|e| {
        Value::error(Names::e_eval(), &format!("Syntax error: {}", e))
    })?;
    let mut result = Value::None;
    for ast in &program {
        result = eval(env, ast)?;
    }
    Ok(result)
}
